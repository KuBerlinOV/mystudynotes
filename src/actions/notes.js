import { v4 as uuidv4 } from 'uuid';

//ADD_NOTE
export const addNote = ({ topic = '', // in this case you should imagine destructuring as you are passing some object with the same data like ' const movie = {title, year, id} '
    description = '',           //when we are passing it to the action function as an argument and destructuring it. 
    note = '',                 //when it would be like ' const {title, year, id} = movie ' 
    reference = '',            //Furthermore, we provide here the default values for the object keys so
    tag = '#',
    status = 'in progress',     //the user could leave these options empty if he wanted to.
    createdAt = 0
} = {}) => {
    return {
        type: 'ADD_NOTE',
        note: {
            id: uuidv4(),
            topic,
            description,
            note,
            reference,
            tag,
            status,
            createdAt
        }
    }
}

//EDIT_NOTE

export const updateNote = (id, updates) => ({
    type: 'UPDATE_NOTE',
    id,
    updates
})

export const updateStatus = (id) => ({
    type: 'UPDATE_STATUS',
    id
})

//REMOVE_NOTE

export const removeNote = (id) => ({
    type: 'REMOVE_NOTE',
    id
})