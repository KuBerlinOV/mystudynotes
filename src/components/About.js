import React from 'react';

const About = () => {
    return (
        <div>organizing your studies</div>
    )
}

export default About;