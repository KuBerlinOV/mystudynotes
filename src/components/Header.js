import React from 'react';
import { NavLink } from 'react-router-dom';


const Header = () => (
    <header>
        <NavLink activeClassName='is-active' className="nav-link" to="/" exact={true}>Home</NavLink>
        <NavLink activeClassName='is-active' className="nav-link" to="/notes">Notes</NavLink>
        <NavLink activeClassName='is-active' className="nav-link" to="/about">Organizing your studies</NavLink>
    </header>
)

export default Header;
