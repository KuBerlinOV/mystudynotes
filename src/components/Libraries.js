import React from 'react';

import Notes from './Notes'

const Libraries = () => {
    return (
        <div>
            <ul>
                <Notes />
            </ul>
        </div>
    )
};

export default Libraries