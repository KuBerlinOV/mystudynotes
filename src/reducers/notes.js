//reducers = notes
const notesReducerDefaultState = []

const notesReducer = (state = notesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_NOTE':
            return [...state, action.note];
        case 'UPDATE_NOTE':
            return state.map(note => {
                if (note.id === action.id) {
                    return {
                        ...note,
                        ...action.updates
                    }
                } else {
                    return note
                }
            });
        case 'REMOVE_NOTE':
            return state.filter(({ id }) => id !== action.id);
        case 'UPDATE_STATUS':
            return state.map(note => {
                if (note.id === action.id && note.status === 'in progress') {
                    return {
                        ...note,
                        status: 'mastered'
                    }
                } else if (note.id === action.id && note.status === 'mastered') {
                    return {
                        ...note,
                        status: 'in progress'
                    }
                } else {
                    return {
                        ...note
                    }
                }
            })
        default:
            return state;
    }

}

export default notesReducer;
