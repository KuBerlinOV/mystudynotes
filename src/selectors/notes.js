


const selectNotes = (notes, { text, sortBy, startDate, endDate }) => {
    return notes.filter(note => {
        const descriptionToMatch = note.description.toLowerCase()
        const topicToMatch = note.topic.toLowerCase();
        const textToMatch = text.toLowerCase();

        const textMatch = descriptionToMatch.includes(textToMatch) || topicToMatch.includes(textToMatch) ? true : false;
        const startDateMatch = typeof startDate !== 'number' || note.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || note.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'status') {
            return a.status === 'mastered' ? 1 : -1
        }
    })
}

export default selectNotes;