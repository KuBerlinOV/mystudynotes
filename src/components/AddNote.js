import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { addNote } from '../actions/notes';
import NoteForm from './NoteForm';

export const AddNote = (props) => {

    return (
        <NoteForm
            handleSubmit={(note) => {
                props.addNote(note);
                props.history.push('/notes');
            }}
        />
    )
}

const mapDispatchToProps = (dispatch) => ({
    addNote: (note) => dispatch(addNote(note))
})

export default connect(undefined, mapDispatchToProps)(AddNote);