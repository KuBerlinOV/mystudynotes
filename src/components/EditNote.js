import React from 'react';
import { connect } from 'react-redux';
import NoteForm from './NoteForm';
import { updateNote } from '../actions/notes';


export const EditNote = (props) => {
    return (
        <div>
            <NoteForm
                //note before update
                note={props.note}
                //this is handle submit props
                handleSubmit={(note) => {
                    //dispatch action to edit the expense
                    props.editNote(props.note.id, note);
                    //redirect to dashBoard
                    props.history.push('/notes')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    editNote: (id, note) => dispatch(updateNote(id, note))
})

const mapStateToProps = (state, props) => {
    return {
        note: state.notes.find((note) => note.id == props.match.params.id)
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditNote);