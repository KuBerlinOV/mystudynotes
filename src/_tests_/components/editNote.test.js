import React from 'react';
import { mount, shallow } from 'enzyme';
import { EditNote } from '../../components/EditNote';
import notes from '../fixtures/notes';

let wrapper, editNoteSpy, historySpy;

beforeEach(() => {
    editNoteSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = mount(<EditNote note={notes[0]} editNote={editNoteSpy} history={historySpy} />);
})

describe('EditNote', () => {


    it('should render EditNote correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should handle submit', () => {
        //interestingly I had troubles with passing this test because NoteForm props, which passed the data of the note 
        //from EditNote page where the state is dispatched, was named as 'noteBeforeUpdate' and not just 'note'.
        //for the future is always important to have the similar name
        wrapper.find('NoteForm').prop('handleSubmit')(notes[0])//<- I grabed the handleSubmit props and called it with the note object
        expect(historySpy.push).toHaveBeenLastCalledWith('/notes')
        expect(editNoteSpy).toHaveBeenCalledWith(notes[0].id, notes[0])
    })
})