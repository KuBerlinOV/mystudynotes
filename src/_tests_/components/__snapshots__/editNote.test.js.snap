// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNote should render EditNote correctly 1`] = `
<EditNote
  editNote={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  note={
    Object {
      "createdAt": 0,
      "description": "redux note",
      "id": "1",
      "note": "here is a note",
      "reference": "some reference",
      "status": "mastered",
      "tag": "some tag",
      "topic": "redux",
    }
  }
>
  <div>
    <NoteForm
      handleSubmit={[Function]}
      note={
        Object {
          "createdAt": 0,
          "description": "redux note",
          "id": "1",
          "note": "here is a note",
          "reference": "some reference",
          "status": "mastered",
          "tag": "some tag",
          "topic": "redux",
        }
      }
    >
      <div>
        <form
          action=""
          data-testid="form"
          onSubmit={[Function]}
        >
          <h3>
            Create Note
          </h3>
          <h3>
            Topic:
          </h3>
          <input
            autoFocus={true}
            name="topic"
            onChange={[Function]}
            placeholder="topic"
            type="text"
            value="redux"
          />
          <h3>
            Description
          </h3>
          <input
            id=""
            name="description"
            onChange={[Function]}
            placeholder="description"
            type="text"
            value="redux note"
          />
          <h3>
            Note
          </h3>
          <textarea
            cols="30"
            id=""
            name="note"
            onChange={[Function]}
            placeholder="write your note here"
            rows="10"
            value="here is a note"
          />
          <h3>
            Reference
          </h3>
          <input
            id=""
            name="reference"
            onChange={[Function]}
            placeholder="reference"
            type="text"
            value="some reference"
          />
          <h3>
            Tag
          </h3>
          <input
            name="tag"
            onChange={[Function]}
            placeholder="tag"
            type="text"
            value="some tag"
          />
          <div>
            <button>
              Save
            </button>
          </div>
        </form>
      </div>
    </NoteForm>
  </div>
</EditNote>
`;
